Framework/API/inc/MantidAPI/Algorithm.tcc

```diff
@@ -0,0 +1,138 @@
+#include "MantidAPI/Algorithm.h"
+#include "MantidAPI/IndexProperty.h"
+#include "MantidAPI/WorkspaceProperty.h"
+
+namespace {
+template <typename T1, typename T2>
+void setWorkspaceProperty(Mantid::API::WorkspaceProperty<T1> *wsProp,
+                          const T2 &wksp, const boost::true_type &) {
+  *wsProp = wksp;
+}
+
+template <typename T1, typename T2>
+void setWorkspaceProperty(Mantid::API::WorkspaceProperty<T1> *wsProp,
+                          const T2 &wksp, const boost::false_type &) {
+  wsProp->setValue(wksp);
+}
+} // namespace
+
+namespace Mantid {
+namespace API {
+/** Declare a property which defines the workspace and allowed index types, as
+* well as a property for capturing the indices all at once. This method is
+* only enabled if T is convertible to MatrixWorkspace.
+@param propertyName Name of property which will be reserved
+@param allowedIndexTypes combination of allowed index types. Default
+IndexType::WorkspaceIndex
+@param optional Determines if workspace property is optional. Default
+PropertyMode::Type::Mandatory
+@param lock Determines whether or not the workspace is locked. Default
+LockMode::Type::Lock
+@param doc Property documentation string.
+*/
+template <typename T, typename>
+void Algorithm::declareIndexProperty(const std::string &propertyName,
+                                     const int allowedIndexTypes,
+                                     PropertyMode::Type optional,
+                                     LockMode::Type lock,
+                                     const std::string &doc) {
+  declareProperty(
+      Kernel::make_unique<WorkspaceProperty<T>>(
+          propertyName, "", Kernel::Direction::Input, optional, lock),
+      doc);
+
+  declareProperty(Kernel::make_unique<IndexTypeProperty>(
```

Comment:
@martyngigg Would it be a compromise to use methods to build the names from the property name?